### PR TITLE
added event list setup and metadata model for new VDR

### PR DESCRIPTION
**Pull request by woutslakhorst**
The eventList is required to sort VDR updates based on the transaction that published it.

The new packages will replace the doc and store package since they are to generic in their naming when used in other packages.

Part of #1709

### CODE REVIEW

File: vdr/diddocuments/didstore/event_test.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+		el.insert(event{TXRef: sha1s, LogicalClock: 1})
+
+		require.Len(t, el.Events, 3)
+		assert.Equal(t, sha0s, el.Events[0].TXRef)
+		assert.Equal(t, sha1s, el.Events[1].TXRef)
+		assert.Equal(t, sha2s, el.Events[2].TXRef)
+	})
+}
+
+func TestEventList_updates(t *testing.T) {
```
Comment (gerardsn) [inline vdr/diddocuments/didstore/event_test.go:101]: can't review this without more context

File: vdr/diddocuments/didstore/event.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// eventList is an in-memory representation of an Events shelf entry
+type eventList struct {
+	Events []event `json:"events"`
+}
+
+func (el *eventList) copy() eventList {
+	return *el
+}
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/diddocuments/didstore/event.go:91]: not needed anymore, removed.

File: vdr/diddocuments/didstore/event.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+// insert the event at the correct location
+func (el *eventList) insert(e event) {
+	// 1% case
+	if len(el.Events) == 0 {
+		el.Events = []event{e}
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/diddocuments/didstore/event.go:97]: Ok 1ns faster...

File: vdr/diddocuments/didstore/event_test.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	})
+}
+
+func TestEventList_copy(t *testing.T) {
+	el := eventList{}
+	cp := el.copy()
+
+	el.insert(event{TXRef: sha0s})
+
+	assert.Equal(t, 0, len(cp.Events))
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/diddocuments/didstore/event_test.go:174]: has been removed. Shallow copy was enough, so no more copy method.

File: vdr/diddocuments/didstore/event.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	if last.LogicalClock < e.LogicalClock {
+		return
+	}
+
+	// 0.01% case
+	sort.Stable(el)
+}
+
+// updates returns the latest matching event and a sublist of updates that need to be applied to the latest.
+func (el *eventList) updates(updated eventList) (*event, []event) {
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/diddocuments/didstore/event.go:113]: added example

File: vdr/didstore/event.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	// Created is the transaction creation time, used for sorting
+	Created time.Time `json:"created"`
+	// LogicalClock contains the LC header from the transaction
+	LogicalClock uint32 `json:"lc"`
+	// TXRef contains the TX.Ref of the original transaction. Used for ordering
+	TXRef hash.SHA256Hash `json:"txref"`
+	// DocRef contains the reference to the document shelf. Equals transaction payload hash
+	DocRef hash.SHA256Hash `json:"docref"`
+	// MetaRef contains a reference to the documentMetadata shelf. "DID + version"
+	MetaRef string `json:"metaref"`
```
Comment (reinkrul) [inline vdr/didstore/event.go:41]: confusing, the other `...Ref`s are actual hashes, this isn't

**Comment by reinkrul** [comment]
I'm missing the grand scheme of things, why do we need this?

File: vdr/didstore/event.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+	// 0.01% case
+	sort.Stable(el)
+}
+
+// diff returns the latest matching event and a sublist of events that need to be applied to the latest.
+// Given transaction orderings A: 1->2->4 and B: 1->2->3.
+// A.diff(B) results in 2, [3,4] where 2 is the last common TX (base version of DID document)
+// and 3,4 are updates that have to be applied to the base version as updates.
+func (el *eventList) diff(updated eventList) (*event, []event) {
```
Comment (reinkrul) [inline vdr/didstore/event.go:117]: `store.Add` only adds 1 `did.Document` at the time, so it looks a bit complicated. When we don't add multiple documents at once, you don't need `diff` and can always use `insert`?

**Comment by reinkrul** [comment]
I think we need an introduction of what the problem is, how to (conceptually) solve this and then map this to the code written.

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: vdr/diddocuments/didstore/event.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+func (el *eventList) Len() int {
+	return len(el.Events)
+}
+
+// Less is part of the methods required for sorting
+func (el *eventList) Less(i, j int) bool {
+	left := el.Events[i]
+	right := el.Events[j]
+
+	if left.LogicalClock < right.LogicalClock {
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/diddocuments/didstore/event.go]: I'll rename to `Clock` as in the TXs

File: vdr/didstore/metadata.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+// documentMetadata is like VDR documentMetadata but usable for storage
+type documentMetadata struct {
+	Version int       `json:"version"`
+	Created time.Time `json:"created"`
+	Updated time.Time `json:"updated"`
+	// Hash of DID document bytes. Is equal to payloadHash in network layer.
+	Hash hash.SHA256Hash `json:"hash"`
+	// PreviousHash of the previous version of this DID document
+	PreviousHash *hash.SHA256Hash `json:"previousHash,omitempty"`
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/didstore/metadata.go:36]: I'll change to no casing for now. This part will be removed when the previous hash will also no longer be used from the within the API call.

File: vdr/didstore/event.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+import (
+	"sort"
+	"time"
+
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+	"github.com/nuts-foundation/nuts-node/network/dag"
+)
+
+// event contains the transaction reference and ordering of all DID document diff
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/didstore/event.go:30]: Global find replace also messed stuff up.

File: vdr/didstore/event.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+type event struct {
+	// Created is the transaction creation time, used for sorting
+	Created time.Time `json:"created"`
+	// LogicalClock contains the LC header from the transaction
+	LogicalClock uint32 `json:"lc"`
+	// TXRef contains the TX.Ref of the original transaction. Used for ordering
+	TXRef hash.SHA256Hash `json:"txref"`
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/didstore/event.go:37]: True, but there are 3 references in an event, TX, Doc and Metadata

File: vdr/didstore/event.go (hunk 6 above)
Comment (woutslakhorst) [reply vdr/didstore/event.go:41]: The other two follow the way of the rest of the code. This one follows the previous/current version of the metadata keys.

File: vdr/didstore/event.go (hunk 7 above)
Comment (woutslakhorst) [reply vdr/didstore/event.go:117]: Updates are added one by one through transactions. They may come out of order. Add doesn't care about ordering, it'll fix it for you.

**Review by gerardsn — CHANGES_REQUESTED** [review]
I took a peak at #1709 as well, and it seems there is a race condition possible on the eventList when multiple updates are added at the same time.

**Review by reinkrul — APPROVED** [review]

File: vdr/didstore/event.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+	for i := range el.Events {
+		if !el.Events[i].TXRef.Equals(updated.Events[i].TXRef) {
+			break
+		}
+		eCopy := el.Events[i]
+		lastCommonEvent = &eCopy
+		lastCommonIndex++
+	}
+
+	newList := eventList{Events: append(el.Events[lastCommonIndex:], updated.Events[lastCommonIndex:]...)}
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply vdr/didstore/event.go:133]: Handled by the `Add` method in the new store. It uses:

```
// transactionIndexShelf has the transaction reference as key and a zero byte as value. Used for duplicate detection
transactionIndexShelf = "txRefV2"
```

**Comment by woutslakhorst** [comment]
> This method is only used in store.Add, right after an insert. I think the output of this method should be equal to insertedList[insertIndex-1], insertedList[insertIndex:]. So returning the index from insert should achieve the same?

Could be, but would move part of the logic to the caller. Having it in here gives a nice encapsulation an easier testing.

**Comment by woutslakhorst** [comment]
> I took a peak at #1709 as well, and it seems there is a race condition possible on the eventList when multiple updates are added at the same time.

That would be part of the next PR then, since the `Add` method should handle that.

**Review by gerardsn — APPROVED** [review]

**Pull request merged**